test/Sema/pattern_matching_optional_promotion.swift.gyb

```diff
@@ -0,0 +1,287 @@
+// RUN: %empty-directory(%t)
+// RUN: %gyb %s -o %t/generated.swift
+// RUN: %target-swift-frontend -typecheck -verify %t/generated.swift
+
+// Test that optional promotion works for refutable patterns when matching
+// patterns. In 'switch' statements, additionally check for AST pattern
+// equivalence to show that an implicitly promoted pattern will have the same
+// space projection as the explicit counterpart.
+//
+// Note: We do not test 'catch' statements because patterns there are always
+// matched against a non-optional.
+
+%{
+  contexts = ['do_block', 'closure', 'result_builder_closure']
+  def get_test_case_introducer(context):
+    if context == 'do_block':
+      return 'do'
+    elif context == 'closure':
+      return 'let _: () -> Void ='
+    elif context == 'result_builder_closure':
+      return '@DummyBuilder var x: Void'
```

Comment:
Does anyone have an opinion on removing the result builder context now that constraint generation for closures has been unified? It makes these tests somewhat harder to codify.

Reply:
Fine with me, as long as you add a couple of simple test cases within result builders